app/models/account.rb

```diff
@@ -13,6 +13,26 @@ class Account < ApplicationRecord
 
   validates :account_number, presence: true
 
+  class << self
+    def from_identity_header(identity_header)
+      Account.find_or_create_by(
+        account_number: identity_header.identity['account_number'],
+        is_internal: identity_header.identity['user']['is_internal']
+      )
+    rescue ActiveRecord::RecordNotUnique
+      update_is_internal!(identity_header)
+      retry
+    end
+
+    private
+
+    def update_is_internal!(identity_header)
+      Account.find_or_create_by(
+        account_number: identity_header.identity['account_number']
+      ).update!(is_internal: identity_header.identity['user']['is_internal'])
+    end
+  end
```

Comment:
I think the steps should be:

1. Find the account by account_id alone
2. Update is_internal to whatever is in the header, if needed (AR handles this condition for you)

```rb
account = Account.find_by(account_number: identity_header.identity['account_number'])
account.update!(is_internal: identity_header.identity['user']['is_internal'])
```

Reply:
Yeah, I was considering the performance here, why should we do an insert on any user interaction? :thinking:

Reply:
The snippet provided does not actually update unless the is_internal field is changed. That's what I mean by 'AR handles this condition for you'. You can try it out locally to confirm.